### PR TITLE
[Feat] Add DINO on MMDetection 3.x

### PR DESCRIPTION
**Pull request by Li-Qingyun**
## Motivation

Implement the results in the paper "[DINO: DETR with Improved DeNoising Anchor Boxes for End-to-End Object Detection](https://arxiv.org/abs/2203.03605)" with mmdetection. The [original code](https://github.com/IDEACVR/DINO) has released.

## Remark

For DINO implementation, our procedure includes three steps:

- [x] **Add DINO on mmdetection 2.x**  #8362 (Already finished!)

- [x] **Design refactor scheme, refactor DETR-like models on mmdetection 3.x.**  #8763 (Finished and merged!)

- [x] **Add(refactor) DINO on mmdetection 3.x**  #9149 (Coming soon)

Besides, there are also brother PRs #8544, #9405,  #8533, #9252 of two algorithms (Condition DETR and DAB DETR) developed in parallel.

I'm a student contributor in GitLink Code Camp. And this is my first experience of contributing a PR to the open source community. I'm trying hard to make DINO available in mmdetection as soon as possible.

**If your project allows to use mmdetection2.x, you can refer to #8362 for the version that has been finished but not merged.**

### CODE REVIEW

File: mmdet/models/dense_heads/dino_head.py
Comment (jshilong) [inline mmdet/models/dense_heads/dino_head.py:254]: dn_bbox_pred.new_tensor([],dtype=torch.long)

File: mmdet/models/detectors/dino.py
Comment (jshilong) [inline mmdet/models/detectors/dino.py:361]: maybe we can move this to the layers/transformer/dino_transformer.py

File: mmdet/models/dense_heads/dino_head.py
Comment (jshilong) [inline mmdet/models/dense_heads/dino_head.py:236]: add `_` for all `x_sinlge` methods or do not add, but keep consistent. You can refer to other detectors.

File: mmdet/models/dense_heads/dino_head.py
Comment (jshilong) [inline mmdet/models/dense_heads/dino_head.py:249]: never add hard code `.cuda`  which may cause potential errors, please to the device of other exist tensors.

File: mmdet/models/detectors/dino.py
Comment (jshilong) [inline mmdet/models/detectors/dino.py:166]: with_query_denoising or use_query_denosing to indicate it is a bool tensor.

File: mmdet/models/detectors/dino.py
Comment (jshilong) [inline mmdet/models/detectors/dino.py:240]: create_dn_queries
generator sounds like a module with parameters

File: mmdet/models/detectors/dino.py
Comment (jshilong) [inline mmdet/models/detectors/dino.py:44]: dn_query_generator

File: mmdet/models/detectors/dino.py
Comment (jshilong) [inline mmdet/models/detectors/dino.py:410]: Why not make it as a nn.module and initialize label_enc in this module?

File: mmdet/models/detectors/dino.py
Comment (jshilong) [inline mmdet/models/detectors/dino.py:461]: fix the  .cuda

File: mmdet/models/detectors/dino.py
Comment (jshilong) [inline mmdet/models/detectors/dino.py:514]: fix the to cuda

File: mmdet/models/layers/transformer/dino_transformer.py
Comment (jshilong) [inline mmdet/models/layers/transformer/dino_transformer.py:24]: pos_tensor seems a points or bbox?
if yes, pos_tensor seems confusing

File: mmdet/models/dense_heads/dino_head.py
Comment (Li-Qingyun) [reply mmdet/models/dense_heads/dino_head.py:236]: I search `_single` globally in the repo:
The `loss_by_feat_single` was adopted by `anchor_head.py`, `atss_head.py`, `rtmdet_head.py`, et. al. and only `maskformer` and `mask2former` adopted `_loss_by_feat_single`
No head adopted `get_targets_single`, they used `_get_targets_single`.

I decide to follow the most cases: use `loss_by_feat_single` and `_xxx_single` for others. Is this OK?

File: mmdet/models/dense_heads/dino_head.py
Comment (Li-Qingyun) [reply mmdet/models/dense_heads/dino_head.py:249]: @jshilong OK, thank you for the advice.

Comment (Li-Qingyun) [reply mmdet/models/dense_heads/dino_head.py:249]: @jshilong OK, thank you for the advice.
I modify them as follows:
```Python
# ori
t = torch.range(0, len(gt_labels) - 1).long().cuda()
# new
t = torch.range(0, len(gt_labels) - 1, dtype=torch.long, device=device)

# ori
pos_inds = (torch.tensor(range(num_groups)) * single_pad).long().cuda().unsqueeze(1) + t
# new
pos_inds = torch.range(0, num_groups, dtype=torch.long, device=device) + t
```

File: mmdet/models/detectors/dino.py
Comment (Li-Qingyun) [reply mmdet/models/detectors/dino.py:166]: OK, I decided to use `self.training` as the indicator.

File: mmdet/models/detectors/dino.py
Comment (Li-Qingyun) [reply mmdet/models/detectors/dino.py:410]: Haha, actually I also had the thought recently.

File: mmdet/models/detectors/dino.py
Comment (Li-Qingyun) [reply mmdet/models/detectors/dino.py:461]: @jshilong   OK, the modifications are:
```python
# origin
positive_idx = torch.tensor(range(
    len(boxes))).long().cuda().unsqueeze(0).repeat(num_groups, 1)
positive_idx += (torch.tensor(range(num_groups)) * len(boxes) *
                 2).long().cuda().unsqueeze(1)
# new
positive_idx = torch.range(0, len(bboxes), dtype=torch.long,
                           device=bboxes.device)
positive_idx = positive_idx.unsqueeze(0).repeat(num_groups, 1)
positive_idx += torch.range(0, num_groups, dtype=torch.long,
                            device=bboxes.device) * len(boxes) * 2
```

File: mmdet/models/layers/transformer/dino_transformer.py
Comment (Li-Qingyun) [reply mmdet/models/layers/transformer/dino_transformer.py:24]: @jshilong  Yes, it is also used in Conditional DETR and DAB DETR, what about name it with `positions`?

File: mmdet/models/dense_heads/deformable_detr_head.py
Comment (jshilong) [inline mmdet/models/dense_heads/deformable_detr_head.py:91]: Please remember to discuss this in next meeting

File: mmdet/models/dense_heads/dino_head.py
Comment (jshilong) [inline mmdet/models/dense_heads/dino_head.py:48]: returned seems inappropriate.
 Only when `as_two_stage` is `True` it will be passed in, otherwise it would be "None".

File: mmdet/models/dense_heads/dino_head.py
Comment (jshilong) [inline mmdet/models/dense_heads/dino_head.py:92]: has shape (num_decoder_layers, bs,num_qeury, 4) with the last dimension arranged as (cx, cy, w, h)

File: mmdet/models/dense_heads/deformable_detr_head.py
Comment (jshilong) [reply mmdet/models/dense_heads/deformable_detr_head.py:91]: We just follow the order of the original implementation

File: mmdet/models/dense_heads/dino_head.py
Comment (jshilong) [inline mmdet/models/dense_heads/dino_head.py:43]: what is the difference between `num_query_total` and `num_query`?

File: mmdet/models/dense_heads/dino_head.py
Comment (jshilong) [inline mmdet/models/dense_heads/dino_head.py:46]: num_feat_points

File: mmdet/models/dense_heads/dino_head.py
Comment (jshilong) [inline mmdet/models/dense_heads/dino_head.py:212]: `get_dn_targets` may be better following our naming convention

File: mmdet/models/dense_heads/dino_head.py
Comment (jshilong) [inline mmdet/models/dense_heads/dino_head.py:251]: Just change to `torch.cat(factors)`, because concatenating the first dimension is the default behavior.

File: mmdet/models/dense_heads/dino_head.py
Comment (jshilong) [inline mmdet/models/dense_heads/dino_head.py:269]: targets

File: mmdet/models/dense_heads/dino_head.py
Comment (jshilong) [inline mmdet/models/dense_heads/dino_head.py:307]: targets

File: mmdet/models/dense_heads/dino_head.py
Comment (jshilong) [inline mmdet/models/dense_heads/dino_head.py:335]: num_query_each_group

File: mmdet/models/detectors/dino.py
Comment (jshilong) [inline mmdet/models/detectors/dino.py:150]: tuple[dict]

File: mmdet/models/detectors/dino.py
Comment (jshilong) [inline mmdet/models/detectors/dino.py:176]: remove all unnecessary comments

File: mmdet/models/detectors/dino.py
Comment (jshilong) [inline mmdet/models/detectors/dino.py:189]: I am confused about this comment, there is no `init_reference_out` here

File: mmdet/models/layers/transformer/dino_transformer.py
Comment (jshilong) [inline mmdet/models/layers/transformer/dino_transformer.py:28]: encoding instead of embedding

**Comment by jshilong** [comment]
The overall design LGTM except for some naming conventions.
Besides I recommend moving the static method `gen_s**` to utils.py.

File: mmdet/models/dense_heads/dino_head.py
Comment (Li-Qingyun) [reply mmdet/models/dense_heads/dino_head.py:43]: `num_query_total = num_denoising_query + num_matching_query`, i think the `num_query` is more likely to be understood as the `num_matching_query`. Hence, the `total` was added to tell it is the total number of the two kinds of queries, which eliminates ambiguity.

File: mmdet/models/dense_heads/dino_head.py
Comment (jshilong) [reply mmdet/models/dense_heads/dino_head.py]: > @jshilong I search `_single` globally in the repo: The `loss_by_feat_single` was adopted by `anchor_head.py`, `atss_head.py`, `rtmdet_head.py`, et. al. and only `maskformer` and `mask2former` adopted `_loss_by_feat_single` No head adopted `get_targets_single`, they used `_get_targets_single`.
> 
> I decide to follow the most cases: use `loss_by_feat_single` and `_xxx_single` for others. Is this OK?

sounds good to me

File: mmdet/models/dense_heads/dino_head.py
Comment (Li-Qingyun) [reply mmdet/models/dense_heads/dino_head.py:94]: OK, I have added the coordinate format to all the related comments.

**Comment by TinyTigerPan** [comment]
Hi, thanks for your wonderful PR, I want to know when it can be merged to dev3.x and 3.x, and MMDet or the author or someone else can provide the ckpt.

File: mmdet/models/dense_heads/dino_head.py
Comment (jshilong) [inline mmdet/models/dense_heads/dino_head.py:126]: Seems using the loss_by_feat of DETR here
 what is the difference of it between DETR and DeformableDETR?

Comment (jshilong) [reply mmdet/models/dense_heads/dino_head.py:126]: add a comment here on why we adopt the loss_by_feat of DETR instead of DefornableDETR

**Comment by Li-Qingyun** [comment]
> Hi, thanks for your wonderful PR, I want to know when it can be merged to dev3.x and 3.x, and MMDet or the author or someone else can provide the ckpt.

Hi! Thanks for your attention. The main coding work has been finished and the training mAP of DINO_4scale_12e (49.0) has been alighed with the results reported in the original paper. We are going on the final reviews and training experiments of other settings. I have only ckpt of DINO_4scale_12e. The other experiments are in squeue. But I have the weights converted from the official repo that may be useful. So which setting do you require.

**Comment by TinyTigerPan** [comment]
> > Hi, thanks for your wonderful PR, I want to know when it can be merged to dev3.x and 3.x, and MMDet or the author or someone else can provide the ckpt.
> 
> Hi! Thanks for your attention. The main coding work has been finished and the training mAP of DINO_4scale_12e (49.0) has been alighed with the results reported in the original paper. We are going on the final reviews and training experiments of other settings. I have only ckpt of DINO_4scale_12e. The other experiments are in squeue. But I have the weights converted from the official repo that may be useful. So which setting do you require.

Thanks! I can wait for the PR merge to dev3.x.

**Comment by Li-Qingyun** [comment]
@TinyTigerPan OK, I'll make it availiable as soon as possible.

**Comment by jshilong** [comment]
1. Please merge the refacor-detr after #9460 is merged then change the data flow to BNC
2. please remove all 91 classes related codes

**Comment by Li-Qingyun** [comment]
> 1. Please merge the refacor-detr after [[Refactor] Change to BNC data flow for all DETRs #9460](https://github.com/open-mmlab/mmdetection/pull/9460) is merged then change the data flow to BNC
> 2. please remove all 91 classes related codes

The merge has been finished. The inference mAP of DINO_4s_12e on the new codebase has been aligned.  @jshilong

File: mmdet/models/layers/transformer/utils.py
Comment (jshilong) [inline mmdet/models/layers/transformer/utils.py:47]: remember to remove this after conditional and dab merged

Comment (Li-Qingyun) [reply mmdet/models/layers/transformer/utils.py:47]: @jshilong okkkk

**Review by jshilong — APPROVED** [review]
LGTM

**Review by RangiLyu — APPROVED** [review]

**Comment by Li-Qingyun** [comment]
A `DINO-5scale-R50` config
```
_base_ = './dino-4scale_r50_8xb2-12e_coco.py'
model = dict(
    backbone=dict(out_indices=(0, 1, 2, 3)),
    neck=dict(in_channels=[256, 512, 1024, 2048], num_outs=5),
    encoder=dict(layer_cfg=dict(self_attn_cfg=dict(num_levels=5))),
    decoder=dict(layer_cfg=dict(cross_attn_cfg=dict(num_levels=5))))
train_dataloader = dict(batch_size=1)

```

**Pull request merged**
